### DISCUSSION TITLE
Is Dynamic/Per-item Colorization for Suggestion Menu Items Possible?

### DISCUSSION THREAD

**Discussion by astridos2go** [nushell/nushell#12581] [Q&A] [answered]
Is there any way to control the colors of specific items a custom completer suggests? 

## Background
I have `zoxide` configured in my shell. I have created a completer that works _almost_ identical to the default `cd` command's completer. (I plan on adding in the additional `zoxide` suggestions later... I'm not insane)

## Example
When I tab complete the builtin `cd` command (`_cd` on my system), my suggestions use some predefined color (I'm actually not sure where this is coming from... Maybe `LS_COLORS`?) 

![image](https://github.com/nushell/nushell/assets/45048997/04d68d5a-d8a3-43f9-a53b-af96f40b716e)

When I use my custom completer for `zoxide` (named `cd`), it uses the colors defined in the `completion_menu` object in the config menus section. 

![image](https://github.com/nushell/nushell/assets/45048997/092a04f2-c630-4049-a7ec-b1e5940889e7)

## Question
My question is, is there any way to define the highlight color for items in the custom completion menu? I know there's the ability to add a description, but can I specify the `shape` that the item is? Maybe pass it a hex code? Could I wrap my items in `(ansi)` blocks?

## Use cases
My use case for this would be when I add in the `zoxide` completions alongside the `cd`-style completions, I'd like to have them in different colors, but could be useful for other completions as well. 

Any ideas would be appreciated! Thanks!

**Comment by fdncred** [comment, marked as answer]
There's been some work around this but it may be unfinished.

https://github.com/nushell/nushell/pull/11442
https://github.com/nushell/nushell/issues/5292
https://github.com/nushell/reedline/pull/691
https://github.com/nushell/reedline/pull/692
https://github.com/nushell/nushell/pull/11444

**Comment by astridos2go** [comment]
Adding a `style` column to the table returned from my custom completion seems to work in nushell 0.92.2! Thanks!

**Reply by fdncred** [reply]
Nice. It would be good to see your example if you can share.

**Reply by astridos2go** [reply]
Sure! 

> [!NOTE]
> I am aware that this is an ugly implementation of what I am trying to achieve. And that it doesn't handle symlinks, nor does it use the `ls_colors` as I would like. It's a work in progress!

```nushell {highlight =10}
def dir-suggester [path_str:string] {
    let path = ($path_str | str trim --char "`" | str trim --char "\\" --right)
    let directory = (
        if ($path | path exists) {$path} 
        else if ($path | path dirname | path exists) {$path | path dirname} 
        else {"."}
    )
    ls $directory --all
    | where type == "dir" 
    | where ($it.name | str starts-with --ignore-case $path)
    | sort-by name --natural --ignore-case
    | select name 
    # Basically, I just add another column with the `style` column, and provide it the same structure that you do in `color_config`
    | par-each --keep-order {|row| [[value style]; [($row.name | dir-escape) { fg: blue, attr: b }]]}
    | flatten
}
```

The function returns this table! Reedline pays attention to the `style` column per the latest release, and overrides the default styling. `carapace` also adds this `style` column now, albeit probably better than I'm doing. 

```
╭───┬─────────────┬─────────────────╮
│ # │    value    │      style      │
├───┼─────────────┼─────────────────┤
│ 0 │ carapace\   │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 1 │ chezmoi\    │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 2 │ ferium\     │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 3 │ glab-cli\   │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 4 │ mypy\       │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 5 │ nushell\    │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 6 │ oh-my-posh\ │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 7 │ op\         │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
│ 8 │ scoop\      │ ╭──────┬──────╮ │
│   │             │ │ fg   │ blue │ │
│   │             │ │ attr │ b    │ │
│   │             │ ╰──────┴──────╯ │
╰───┴─────────────┴─────────────────╯
```
Hope this is helpful, there's probably a simpler example that could be made.
